exchanges/stream/processor.go

```diff
@@ -0,0 +1,117 @@
+	"github.com/thrasher-corp/gocryptotrader/exchanges/asset"
+)
+
+const (
+	// Book is a key type for websocket orderbook data processing
+	Book                     UpdateType = iota
+	defaultChannelBufferSize            = 10
+)
+
+var (
+	errMaxChanBufferSizeInvalid = errors.New("max channel buffer must be greater than 0")
+	errDataHandlerMustNotBeNil  = errors.New("data handler cannot be nil")
+	errChannelFull              = errors.New("channel full")
+	errKeyEmpty                 = errors.New("key is empty")
+	errNoFunctionalityToProcess = errors.New("no functionality to process")
+)
+
+// Processor is a stream processor that handles incoming data from a stream,
+// spawns workers for each unique key, and processes the data in a concurrent
+// manner.
+type Processor struct {
+	routes         map[Key]chan func() error
+	chanBufferSize int
+	dataHandler    chan interface{}
+	mtx            sync.RWMutex
+}
+
+// Key is a unique key for an individual type of work.
+type Key struct {
+	Type  UpdateType
+	Asset asset.Item
+	Base  *currency.Item
+	Quote *currency.Item
+}
+
+// UpdateType is a unique key for an individual type of work.
+type UpdateType uint8
+
+// NewProcessor returns a new stream processor. This handles incoming data
+// from a stream and processes it in a concurrent manner. It will spawn a
+// new worker for each unique key. If a worker already exists for a key, it
+// will use that worker. Functions are queued and processed in the order
+// they are received. Function closures are used for ease of implementation
+// with current systems.
+func NewProcessor(maxChanBuffer int, dataHandler chan interface{}) (*Processor, error) {
+	if maxChanBuffer <= 0 {
+		return nil, errMaxChanBufferSizeInvalid
+	}
+	if dataHandler == nil {
+		return nil, errDataHandlerMustNotBeNil
+	}
+	return &Processor{
+		routes:         make(map[Key]chan func() error),
+		chanBufferSize: maxChanBuffer,
+		dataHandler:    dataHandler,
+	}, nil
+}
+
+// Process spawns a new worker for a key if one does not already exist. It
+// will then queue the function to be processed. If the channel is full, it
+// will block until a slot is available. This tries to alleviate websocket
+// reader blocking issues.
+func (w *Processor) Process(key Key, fn func() error) error {
+	if key == (Key{}) {
+		return errKeyEmpty
+	}
+	if fn == nil {
+		return fmt.Errorf("%w for %v", errNoFunctionalityToProcess, key)
+	}
+
+	w.mtx.RLock()
+	ch, ok := w.routes[key]
+	if !ok {
+		wait := make(chan struct{})
+		go func() {
+			w.mtx.Lock()
+			// re-check under main lock to ensure we don't spawn multiple
+			// workers.
+			if ch, ok = w.routes[key]; ok {
+				close(wait)
+				w.mtx.Unlock()
+				return // already spawned
+			}
+
+			ch = make(chan func() error, w.chanBufferSize)
+			w.routes[key] = ch
+			close(wait)
+			w.mtx.Unlock()
+			for processWebsocketTask := range ch {
+				err := processWebsocketTask()
+				if err != nil {
+					w.dataHandler <- err
+				}
+			}
```

Comment:
This comment is a bit theoretical because its making assumptions about future uses. As I said last time, I actually quite like having dedicated routes to process messages. 

Depending on how this goes moving forward  I anticipate two problems with this design:
**1)** We have a system to allow the overriding of the DataHandler to allow a user to have custom processes for how they wish to handle WS updates. This system circumvents that. For the specific use-case you've done here with orderbooks, it doesn't apply because we use `LoadSnapshot(...)` and `Update(...)` which already circumvent the datahandler. 
If you expand this to tickers or orders, we process those via the DataHandler. 
I would no longer have the ability to process these updates and means that it can only be done via GCT and requires an update to `main` GCT to change. Which leads to the next part

**2)** That it can't be toggled. Depending on this feature and how it is going to be accomplished, I think this is something that should be a feature and if not enabled, sends data to the datahandler. 
This all would require a documentation update in terms of having a readme.md for it.

This isn't something to really address in this PR though, because its not being utilised in this way. But you've clearly designed it in such a way as to allow for future expansion. If this is only going to stay to help process orderbooks asynchronously, it could potentially be renamed to be more specific to the orderbook.

Thank you for listening to my TED talk

Reply:
For point 1), I understand your concern about the ability to override the DataHandler for custom processing. However, I'd like to clarify that for tickers as an example, we perform pre-processing within the WebSocket handle data function. This includes reading data from the connection, unmarshalling the byte array into our declared type, and allocating memory for a ticker.Price struct. With the changes proposed, we are essentially moving this unmarshalling and struct allocation process to a dedicated routine. The key point here is that the processed data will still be pushed directly to the DataHandler in all cases. Therefore, while it may seem like a shift, it doesn't bypass the DataHandler's processing within the WebSocket routine manager.

Regarding point 2), Toggling is a good point, If for some reason you want to read and processing linearly then I can see this as acceptable.

Reply:
> With the changes proposed, we are essentially moving this unmarshalling and struct allocation process to a dedicated routine

Does that mean you will be [moving the json unmarshalling as part of the processor.Process](https://github.com/thrasher-corp/gocryptotrader/pull/1316/files#diff-c737e2208dda52b4019818edc7d9c8ccc622b1e6a230baa26c4a1cadd455e833R281)? 

The other part of what you say regarding ticker is only theoretical because its the DataHandler that processes tickers at the moment, adding this to it would only be adding an additional async queue to then process the ticker in a sync channel on the data handler - but this is all theoretical because the work for tickers doesn't exist yet. If you're saying that everything will continue to be pushed to the sync data handler, then there's no real reason for a enable/disable, its just a discussion on what your intent is with this design.

Reply:
> Does that mean you will be [moving the json unmarshalling as part of the processor.Process]

I need the data for the key so I cannot under this design. 

> adding this to it would only be adding an additional async queue to then process the ticker in a sync channel on the data handler 

Yeah a cost benefit would need to be figured out in this case.